[package] googleapis/cci.*: missing dependencies between googleapis libraries

### Description

The `googleapis` package includes multiple libraries, but does not have the dependencies between these libraries.  AFAICT, the generated rules to compile these libraries include these dependencies, but the generated `Findgoogleapis*` files do not preserve these dependencies.

I can try to prepare a PR to solve this problem.  It seems that it would require defining a component for each library, with the right dependencies for each.

### Package and Environment Details

* Package Name/Version: **googleapis/cci.20220711**
* Operating System+version: **Linux**
* Compiler+version: **GCC 12**
* Conan version: **conan 1.55.0**
* Python version: **Python 3.10.9**


### Conan profile

Configuration for profile default:

[settings]
os=Linux
os_build=Linux
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=12
compiler.libcxx=libstdc++
build_type=Release
[options]
[conf]
[build_requires]
[env]


### Steps to reproduce

Apply the following patch and then test with `conan test test_package googleapis/cci.20220711@`

```patch
diff --git a/recipes/googleapis/all/test_package/test_package.cpp b/recipes/googleapis/all/test_package/test_package.cpp
index 32f0b086e..3b9a08aef 100644
--- a/recipes/googleapis/all/test_package/test_package.cpp
+++ b/recipes/googleapis/all/test_package/test_package.cpp
@@ -1,14 +1,13 @@
 #include <iostream>
-#include "google/type/color.pb.h"
+#include <google/bigtable/v2/bigtable.pb.h>
 
 int main() {
     std::cout << "Conan - test package for googleapis\n";
 
-    google::type::Color c;
-    c.set_red(255);
-    c.set_blue(255);
+    google::bigtable::v2::CheckAndMutateRowRequest request;
+    request.set_table_name("projects/my-project/instances/my-instance/tables/my-table");
 
-    std::cout << c.DebugString();
+    std::cout << request.DebugString();
 
     return 0;
 }
```


### Logs

<details><summary>Click to expand log</summary>

```
conan test test_package googleapis/cci.20220711@
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=12
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

googleapis/cci.20220711 (test package): Installing package
Requirements
    googleapis/cci.20220711 from 'conancenter' - Cache
    protobuf/3.21.4 from 'conancenter' - Cache
    zlib/1.2.13 from 'conancenter' - Cache
Packages
    googleapis/cci.20220711:1f2fc6cba1a905f299dfce1a8bdf819ef1ecb710 - Cache
    protobuf/3.21.4:d403f0cadc6326871b16f4359bea9431daa4a81a - Cache
    zlib/1.2.13:2a19826344ff00be1c04403f2f8e7008ed3a7cc6 - Cache

Installing (downloading, building) binaries...
zlib/1.2.13: Already installed!
protobuf/3.21.4: Already installed!
protobuf/3.21.4: Appending PATH environment variable: /usr/local/google/home/coryan/.conan/data/protobuf/3.21.4/_/_/package/d403f0cadc6326871b16f4359bea9431daa4a81a/bin
googleapis/cci.20220711: Already installed!
googleapis/cci.20220711 (test package): Generator 'VirtualRunEnv' calling 'generate()'
googleapis/cci.20220711 (test package): Generator txt created conanbuildinfo.txt
googleapis/cci.20220711 (test package): WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
googleapis/cci.20220711 (test package): Generator 'CMakeDeps' calling 'generate()'
googleapis/cci.20220711 (test package): Calling generate()
googleapis/cci.20220711 (test package): WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
googleapis/cci.20220711 (test package): Preset 'release' added to CMakePresets.json. Invoke it manually using 'cmake --preset release'
googleapis/cci.20220711 (test package): If your CMake version is not compatible with CMakePresets (<3.19) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/usr/local/google/home/coryan/cci-develop/recipes/googleapis/all/test_package/build/generators/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
googleapis/cci.20220711 (test package): Aggregating env generators
googleapis/cci.20220711 (test package): Generated conaninfo.txt
googleapis/cci.20220711 (test package): Generated graphinfo
Using lockfile: '/usr/local/google/home/coryan/cci-develop/recipes/googleapis/all/test_package/build/generators/conan.lock'
Using cached profile from lockfile
googleapis/cci.20220711 (test package): Calling build()
googleapis/cci.20220711 (test package): CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/usr/local/google/home/coryan/cci-develop/recipes/googleapis/all/test_package/build/generators/conan_toolchain.cmake" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/usr/local/google/home/coryan/cci-develop/recipes/googleapis/all/test_package/."
-- Using Conan toolchain: /usr/local/google/home/coryan/cci-develop/recipes/googleapis/all/test_package/build/generators/conan_toolchain.cmake
-- Conan: Target declared 'googleapis::googleapis'
-- Conan: Component target declared 'protobuf::libprotobuf'
-- Conan: Component target declared 'protobuf::libprotoc'
-- Conan: Target declared 'protobuf::protobuf'
-- Conan: Target declared 'ZLIB::ZLIB'
-- Conan: Including build module from '/usr/local/google/home/coryan/.conan/data/protobuf/3.21.4/_/_/package/d403f0cadc6326871b16f4359bea9431daa4a81a/lib/cmake/protobuf/protobuf-generate.cmake'
-- Conan: Including build module from '/usr/local/google/home/coryan/.conan/data/protobuf/3.21.4/_/_/package/d403f0cadc6326871b16f4359bea9431daa4a81a/lib/cmake/protobuf/protobuf-module.cmake'
-- Conan: Including build module from '/usr/local/google/home/coryan/.conan/data/protobuf/3.21.4/_/_/package/d403f0cadc6326871b16f4359bea9431daa4a81a/lib/cmake/protobuf/protobuf-options.cmake'
-- Configuring done
-- Generating done
-- Build files have been written to: /usr/local/google/home/coryan/cci-develop/recipes/googleapis/all/test_package/build/Release
googleapis/cci.20220711 (test package): CMake command: cmake --build "/usr/local/google/home/coryan/cci-develop/recipes/googleapis/all/test_package/build/Release" '--' '-j96'
Consolidate compiler generated dependencies of target test_package
[ 50%] Building CXX object CMakeFiles/test_package.dir/test_package.cpp.o
[100%] Linking CXX executable test_package
/usr/bin/ld: /usr/local/google/home/coryan/.conan/data/googleapis/cci.20220711/_/_/package/1f2fc6cba1a905f299dfce1a8bdf819ef1ecb710/lib/libgoogle_bigtable_v2_bigtable_proto.a(bigtable.pb.cc.o):(.data.rel.ro+0x120): undefined reference to `descriptor_table_google_2fapi_2fannotations_2eproto'
/usr/bin/ld: /usr/local/google/home/coryan/.conan/data/googleapis/cci.20220711/_/_/package/1f2fc6cba1a905f299dfce1a8bdf819ef1ecb710/lib/libgoogle_bigtable_v2_bigtable_proto.a(bigtable.pb.cc.o):(.data.rel.ro+0x128): undefined reference to `descriptor_table_google_2fapi_2fclient_2eproto'
/usr/bin/ld: /usr/local/google/home/coryan/.conan/data/googleapis/cci.20220711/_/_/package/1f2fc6cba1a905f299dfce1a8bdf819ef1ecb710/lib/libgoogle_bigtable_v2_bigtable_proto.a(bigtable.pb.cc.o):(.data.rel.ro+0x130): undefined reference to `descriptor_table_google_2fapi_2ffield_5fbehavior_2eproto'
/usr/bin/ld: /usr/local/google/home/coryan/.conan/data/googleapis/cci.20220711/_/_/package/1f2fc6cba1a905f299dfce1a8bdf819ef1ecb710/lib/libgoogle_bigtable_v2_bigtable_proto.a(bigtable.pb.cc.o):(.data.rel.ro+0x138): undefined reference to `descriptor_table_google_2fapi_2fresource_2eproto'
/usr/bin/ld: /usr/local/google/home/coryan/.conan/data/googleapis/cci.20220711/_/_/package/1f2fc6cba1a905f299dfce1a8bdf819ef1ecb710/lib/libgoogle_bigtable_v2_bigtable_proto.a(bigtable.pb.cc.o):(.data.rel.ro+0x140): undefined reference to `descriptor_table_google_2fapi_2frouting_2eproto'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/test_package.dir/build.make:395: test_package] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/test_package.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2```

</details>

--- a/recipes/googleapis/all/CMakeLists.txt
+++ b/recipes/googleapis/all/CMakeLists.txt
@@ -9,3 +9,5 @@ conan_basic_setup(TARGETS)
 find_package(Protobuf REQUIRED CONFIG)
 
 set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
+
+include("${CMAKE_CURRENT_SOURCE_DIR}/generated_targets.cmake")

--- a/recipes/googleapis/all/conandata.yml
+++ b/recipes/googleapis/all/conandata.yml
@@ -22,9 +22,17 @@ sources:
 patches:
   "cci.20220711":
     - patch_file: "patches/cci.20220711/001-fix-google-api-deps.patch"
+      patch_description: "Fix incorrect dependency name in BUILD.bazel file"
+      patch_type: "bugfix"
   "cci.20220531":
     - patch_file: "patches/cci.20220531/001-fix-google-api-deps.patch"
+      patch_description: "Fix incorrect dependency name in BUILD.bazel file"
+      patch_type: "bugfix"
   "cci.20211122":
     - patch_file: "patches/cci.20211122/001-fix-google-api-deps.patch"
+      patch_description: "Fix incorrect dependency name in BUILD.bazel file"
+      patch_type: "bugfix"
   "cci.20210730":
     - patch_file: "patches/cci.20210730/001-fix-google-api-deps.patch"
+      patch_description: "Fix incorrect dependency name in BUILD.bazel file"
+      patch_type: "bugfix"

--- a/recipes/googleapis/all/conandata.yml
+++ b/recipes/googleapis/all/conandata.yml
@@ -24,15 +24,19 @@ patches:
     - patch_file: "patches/cci.20220711/001-fix-google-api-deps.patch"
       patch_description: "Fix incorrect dependency name in BUILD.bazel file"
       patch_type: "bugfix"
+      patch_source: "https://github.com/googleapis/googleapis/commit/5fd35b6a1316570df7f117426ed35af9086e9bda"
   "cci.20220531":
     - patch_file: "patches/cci.20220531/001-fix-google-api-deps.patch"
       patch_description: "Fix incorrect dependency name in BUILD.bazel file"
       patch_type: "bugfix"
+      patch_source: "https://github.com/googleapis/googleapis/commit/5fd35b6a1316570df7f117426ed35af9086e9bda"
   "cci.20211122":
     - patch_file: "patches/cci.20211122/001-fix-google-api-deps.patch"
       patch_description: "Fix incorrect dependency name in BUILD.bazel file"
       patch_type: "bugfix"
+      patch_source: "https://github.com/googleapis/googleapis/commit/5fd35b6a1316570df7f117426ed35af9086e9bda"
   "cci.20210730":
     - patch_file: "patches/cci.20210730/001-fix-google-api-deps.patch"
       patch_description: "Fix incorrect dependency name in BUILD.bazel file"
       patch_type: "bugfix"
+      patch_source: "https://github.com/googleapis/googleapis/commit/5fd35b6a1316570df7f117426ed35af9086e9bda"

--- a/recipes/googleapis/all/conandata.yml
+++ b/recipes/googleapis/all/conandata.yml
@@ -19,3 +19,12 @@ sources:
   "cci.20210730":  # Used by grpc 1.46.3
     url: "https://github.com/googleapis/googleapis/archive/2f9af297c84c55c8b871ba4495e01ade42476c92.zip"
     sha256: "c53ef0768e07bd4e2334cdacba8c6672d2252bef307a889f94893652e8e7f3a4"
+patches:
+  "cci.20220711":
+    - patch_file: "patches/cci.20220711/001-fix-google-api-deps.patch"
+  "cci.20220531":
+    - patch_file: "patches/cci.20220531/001-fix-google-api-deps.patch"
+  "cci.20211122":
+    - patch_file: "patches/cci.20211122/001-fix-google-api-deps.patch"
+  "cci.20210730":
+    - patch_file: "patches/cci.20210730/001-fix-google-api-deps.patch"

--- a/recipes/googleapis/all/conanfile.py
+++ b/recipes/googleapis/all/conanfile.py
@@ -71,7 +71,7 @@ class GoogleAPIS(ConanFile):
     def build_requirements(self):
         self.build_requires('protobuf/3.21.4')
         # CMake >= 3.20 is required. There is a proto with dots in the name 'k8s.min.proto' and CMake fails to generate project files
-        self.build_requires('cmake/[>3.19]')
+        self.build_requires('cmake/3.20')
 
     @functools.lru_cache(1)
     def _configure_cmake(self):

--- a/recipes/googleapis/all/conanfile.py
+++ b/recipes/googleapis/all/conanfile.py
@@ -1,7 +1,6 @@
 import os
 import functools
 import glob
-from dataclasses import dataclass
 
 from conans import CMake, tools
 
@@ -15,12 +14,6 @@ from conan.errors import ConanInvalidConfiguration
 from helpers import parse_proto_libraries
 
 required_conan_version = ">=1.45.0"
-
-
-@dataclass
-class Node:
-    mark: str
-    deps: list
 
 
 class GoogleAPIS(ConanFile):
@@ -180,38 +173,10 @@ class GoogleAPIS(ConanFile):
         copy(self, pattern="*.dylib", src=self.build_folder, dst=os.path.join(self.package_folder, "lib"), keep_path=False)
         copy(self, pattern="*.a", src=self.build_folder, dst=os.path.join(self.package_folder, "lib"), keep_path=False)
 
-        graph = {}
-        types = {}
-        for lib in filter(lambda u: u.is_used, self._parse_proto_libraries()):
-            graph[lib.cmake_target] = Node(mark=None, deps=lib.cmake_deps)
-            types[lib.cmake_target] = 'LIB' if lib.srcs else 'INTERFACE'
-
-        def visit(name : str, dag : dict, L : list):
-            # Ignore external deps that are not in the dag, such as protobuf::libprotobuf.
-            # Also skip nodes that are already completely processed
-            if name not in dag or dag[name].mark == 'DONE':
-                return
-            # This would indicate a cycle in the `googleapis` components. The Bazel-based
-            # builds would have failed if such a cycle existed.
-            assert dag[name].mark != 'InProgress'
-            dag[name].mark = 'InProgress'
-            for m in dag[name].deps:
-                visit(m, dag, L)
-            dag[name].mark = 'DONE'
-            L.insert(0, name)
-
-        def tsort():
-            result = []
-            for name in graph:
-                visit(name, graph, result)
-            return result
-
-        ts = tsort()
         with open(os.path.join(self.package_folder, self._DEPS_FILE), "w", encoding="utf-8") as f:
-            for name in ts:
-                if name not in graph:
-                    continue
-                f.write(f"{name} {types[name]} {','.join(graph[name].deps)}\n")
+            for lib in filter(lambda u: u.is_used, self._parse_proto_libraries()):
+                interface = 'LIB' if lib.srcs else 'INTERFACE'
+                f.write(f"{lib.cmake_target} {interface} {','.join(lib.cmake_deps)}\n")
 
     def package_id(self):
         self.info.requires["protobuf"].full_package_mode()

--- a/recipes/googleapis/all/conanfile.py
+++ b/recipes/googleapis/all/conanfile.py
@@ -1,6 +1,7 @@
 import os
 import functools
 import glob
+from io import StringIO
 
 from conans import CMake, tools
 
@@ -65,13 +66,28 @@ class GoogleAPIS(ConanFile):
         if self.options.shared and not self.options["protobuf"].shared:
             raise ConanInvalidConfiguration("If built as shared, protobuf must be shared as well. Please, use `protobuf:shared=True`")
 
+    @property
+    def _cmake_new_enough(self):
+        try:
+            import re
+            output = StringIO()
+            self.run("cmake --version", output=output)
+            m = re.search(r'cmake version (\d+)\.(\d+)\.(\d+)', output.getvalue())
+            major, minor = int(m.group(1)), int(m.group(2))
+            assert major >= 3 and minor >= 20
+        except:
+            return False
+        else:
+            return True
+
     def requirements(self):
         self.requires('protobuf/3.21.4')
 
     def build_requirements(self):
         self.build_requires('protobuf/3.21.4')
         # CMake >= 3.20 is required. There is a proto with dots in the name 'k8s.min.proto' and CMake fails to generate project files
-        self.build_requires('cmake/3.20')
+        if not self._cmake_new_enough:
+            self.build_requires('cmake/3.23.2')
 
     @functools.lru_cache(1)
     def _configure_cmake(self):

--- a/recipes/googleapis/all/conanfile.py
+++ b/recipes/googleapis/all/conanfile.py
@@ -190,3 +190,5 @@ class GoogleAPIS(ConanFile):
                 if libtype == 'LIB':
                     self.cpp_info.components[name].libs = [name]
                 self.cpp_info.components[name].names["pkg_config"] = name
+                if self.settings.os == "Linux":
+                    self.cpp_info.components[name].system_libs.extend(["m"])

--- a/recipes/googleapis/all/conanfile.py
+++ b/recipes/googleapis/all/conanfile.py
@@ -1,6 +1,7 @@
 import os
 import functools
 import glob
+from dataclasses import dataclass
 
 from conans import CMake, tools
 
@@ -14,6 +15,13 @@ from conan.errors import ConanInvalidConfiguration
 from helpers import parse_proto_libraries
 
 required_conan_version = ">=1.45.0"
+
+
+@dataclass
+class Node:
+    mark: str
+    deps: list
+
 
 class GoogleAPIS(ConanFile):
     name = "googleapis"
@@ -172,12 +180,6 @@ class GoogleAPIS(ConanFile):
         copy(self, pattern="*.dylib", src=self.build_folder, dst=os.path.join(self.package_folder, "lib"), keep_path=False)
         copy(self, pattern="*.a", src=self.build_folder, dst=os.path.join(self.package_folder, "lib"), keep_path=False)
 
-        from dataclasses import dataclass
-        @dataclass
-        class Node:
-            mark: str
-            deps: list()
-
         graph = {}
         types = {}
         for lib in filter(lambda u: u.is_used, self._parse_proto_libraries()):
@@ -199,11 +201,11 @@ class GoogleAPIS(ConanFile):
             L.insert(0, name)
 
         def tsort():
-            sorted = []
-            for name in graph.keys():
-                visit(name, graph, sorted)
-            return sorted
-        
+            result = []
+            for name in graph:
+                visit(name, graph, result)
+            return result
+
         ts = tsort()
         with open(os.path.join(self.package_folder, self._DEPS_FILE), "w", encoding="utf-8") as f:
             for name in ts:

--- a/recipes/googleapis/all/conanfile.py
+++ b/recipes/googleapis/all/conanfile.py
@@ -118,6 +118,7 @@ class GoogleAPIS(ConanFile):
         if Version(self.deps_cpp_info["protobuf"].version) <= "3.21.5" and self.settings.os == "Macos" or \
             self.settings.os == "Android":
             deactivate_library("//google/storagetransfer/v1:storagetransfer_proto")
+            deactivate_library("//google/storagetransfer/v1:storagetransfer_cc_proto")
         #  - Inconvenient macro names from /usr/include/math.h : DOMAIN
         if (self.settings.os == "Linux" and self.settings.compiler == "clang" and self.settings.compiler.libcxx == "libc++") or \
             self.settings.compiler in ["Visual Studio", "msvc"]:

--- a/recipes/googleapis/all/conanfile.py
+++ b/recipes/googleapis/all/conanfile.py
@@ -1,6 +1,7 @@
 import os
 import functools
 import glob
+import graphlib
 from io import StringIO
 
 from conans import CMake, tools
@@ -38,9 +39,12 @@ class GoogleAPIS(ConanFile):
 
     def export_sources(self):
         self.copy("CMakeLists.txt")
+        for patch in self.conan_data.get("patches", {}).get(self.version, []):
+            self.copy(patch["patch_file"])
 
     def source(self):
         get(self, **self.conan_data["sources"][str(self.version)], destination=self.source_folder, strip_root=True)
+        self._patch_sources()
 
     def config_options(self):
         if self.settings.os == "Windows":
@@ -156,11 +160,20 @@ class GoogleAPIS(ConanFile):
 
     def build(self):
         proto_libraries = self._parse_proto_libraries()
-        with open(os.path.join(self.source_folder, "CMakeLists.txt"), "a", encoding="utf-8") as f:
+        # Use a separate file to host the generated code, which is generated in full each time.
+        # This is safe to call multiple times, for example, if you need to invoke `conan build` more than
+        # once.
+        with open(os.path.join(self.source_folder, "generated_targets.cmake"), "w", encoding="utf-8") as f:
+            f.write("# Generated C++ library targets for googleapis\n")
+            f.write("# DO NOT EDIT - change the generation code in conanfile.py instead\n")
             for it in filter(lambda u: u.is_used, proto_libraries):
                 f.write(it.cmake_content)
         cmake = self._configure_cmake()
         cmake.build()
+
+    def _patch_sources(self):
+        for patch in self.conan_data.get("patches", {}).get(self.version, []):
+            tools.patch(**patch)
 
     def package(self):
         copy(self, pattern="LICENSE", src=self.source_folder, dst=os.path.join(self.package_folder, "licenses"))
@@ -172,12 +185,26 @@ class GoogleAPIS(ConanFile):
         copy(self, pattern="*.so*", src=self.build_folder, dst=os.path.join(self.package_folder, "lib"), keep_path=False)
         copy(self, pattern="*.dylib", src=self.build_folder, dst=os.path.join(self.package_folder, "lib"), keep_path=False)
         copy(self, pattern="*.a", src=self.build_folder, dst=os.path.join(self.package_folder, "lib"), keep_path=False)
+        graph = {}
+        types = {}
+        for lib in filter(lambda u: u.is_used, self._parse_proto_libraries()):
+            graph[lib.cmake_target] = lib.cmake_deps
+            types[lib.cmake_target] = 'LIB' if lib.srcs else 'INTERFACE'
+        ts = graphlib.TopologicalSorter(graph)
+        with open(os.path.join(self.package_folder, "generated_targets.deps"), "w", encoding="utf-8") as f:
+            for name in ts.static_order():
+                if name not in graph:
+                    continue
+                f.write("{} {} {}\n".format(name, types[name], ','.join(graph[name])))
 
     def package_id(self):
         self.info.requires["protobuf"].full_package_mode()
 
     def package_info(self):
-        # We are not creating components, we can just collect the libraries
-        self.cpp_info.libs = tools.collect_libs(self)
-        if self.settings.os == "Linux":
-            self.cpp_info.system_libs.extend(["m"])
+        with open(os.path.join(self.package_folder, "generated_targets.deps"), "r", encoding="utf-8") as f:
+            for line in f.read().splitlines():
+                (name, type, deps) = line.rstrip('\n').split(' ')
+                self.cpp_info.components[name].requires = deps.split(',')
+                if type == 'LIB':
+                    self.cpp_info.components[name].libs = [name]
+                self.cpp_info.components[name].names["pkg_config"] = name

--- a/recipes/googleapis/all/conanfile.py
+++ b/recipes/googleapis/all/conanfile.py
@@ -1,7 +1,6 @@
 import os
 import functools
 import glob
-import graphlib
 
 from conans import CMake, tools
 
@@ -157,7 +156,10 @@ class GoogleAPIS(ConanFile):
 
     def _patch_sources(self):
         for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
+            # Using **patch here results in an error with the required `patch_description` field.
+            tools.patch(patch_file=patch['patch_file'])
+
+    _DEPS_FILE = "res/generated_targets.deps"
 
     def package(self):
         copy(self, pattern="LICENSE", src=self.source_folder, dst=os.path.join(self.package_folder, "licenses"))
@@ -169,23 +171,51 @@ class GoogleAPIS(ConanFile):
         copy(self, pattern="*.so*", src=self.build_folder, dst=os.path.join(self.package_folder, "lib"), keep_path=False)
         copy(self, pattern="*.dylib", src=self.build_folder, dst=os.path.join(self.package_folder, "lib"), keep_path=False)
         copy(self, pattern="*.a", src=self.build_folder, dst=os.path.join(self.package_folder, "lib"), keep_path=False)
+
+        from dataclasses import dataclass
+        @dataclass
+        class Node:
+            mark: str
+            deps: list()
+
         graph = {}
         types = {}
         for lib in filter(lambda u: u.is_used, self._parse_proto_libraries()):
-            graph[lib.cmake_target] = lib.cmake_deps
+            graph[lib.cmake_target] = Node(mark=None, deps=lib.cmake_deps)
             types[lib.cmake_target] = 'LIB' if lib.srcs else 'INTERFACE'
-        ts = graphlib.TopologicalSorter(graph)
-        with open(os.path.join(self.package_folder, "generated_targets.deps"), "w", encoding="utf-8") as f:
-            for name in ts.static_order():
+
+        def visit(name : str, dag : dict, L : list[str]):
+            # Ignore external deps that are not in the dag, such as protobuf::libprotobuf.
+            # Also skip nodes that are already completely processed
+            if name not in dag or dag[name].mark == 'DONE':
+                return
+            # This would indicate a cycle in the `googleapis` components. The Bazel-based
+            # builds would have failed if such a cycle existed.
+            assert dag[name].mark != 'InProgress'
+            dag[name].mark = 'InProgress'
+            for m in dag[name].deps:
+                visit(m, dag, L)
+            dag[name].mark = 'DONE'
+            L.insert(0, name)
+
+        def tsort():
+            sorted = []
+            for name in graph.keys():
+                visit(name, graph, sorted)
+            return sorted
+        
+        ts = tsort()
+        with open(os.path.join(self.package_folder, self._DEPS_FILE), "w", encoding="utf-8") as f:
+            for name in ts:
                 if name not in graph:
                     continue
-                f.write(f"{name} {types[name]} {','.join(graph[name])}\n")
+                f.write(f"{name} {types[name]} {','.join(graph[name].deps)}\n")
 
     def package_id(self):
         self.info.requires["protobuf"].full_package_mode()
 
     def package_info(self):
-        with open(os.path.join(self.package_folder, "generated_targets.deps"), "r", encoding="utf-8") as f:
+        with open(os.path.join(self.package_folder, self._DEPS_FILE), "r", encoding="utf-8") as f:
             for line in f.read().splitlines():
                 (name, libtype, deps) = line.rstrip('\n').split(' ')
                 self.cpp_info.components[name].requires = deps.split(',')

--- a/recipes/googleapis/all/conanfile.py
+++ b/recipes/googleapis/all/conanfile.py
@@ -2,7 +2,6 @@ import os
 import functools
 import glob
 import graphlib
-from io import StringIO
 
 from conans import CMake, tools
 
@@ -69,25 +68,10 @@ class GoogleAPIS(ConanFile):
     def requirements(self):
         self.requires('protobuf/3.21.4')
 
-    @property
-    def _cmake_new_enough(self):
-        try:
-            import re
-            output = StringIO()
-            self.run("cmake --version", output=output)
-            m = re.search(r'cmake version (\d+)\.(\d+)\.(\d+)', output.getvalue())
-            major, minor = int(m.group(1)), int(m.group(2))
-            assert major >= 3 and minor >= 20
-        except:
-            return False
-        else:
-            return True
-
     def build_requirements(self):
         self.build_requires('protobuf/3.21.4')
         # CMake >= 3.20 is required. There is a proto with dots in the name 'k8s.min.proto' and CMake fails to generate project files
-        if not self._cmake_new_enough:
-            self.build_requires('cmake/3.23.2')
+        self.build_requires('cmake/[>3.19]')
 
     @functools.lru_cache(1)
     def _configure_cmake(self):
@@ -195,7 +179,7 @@ class GoogleAPIS(ConanFile):
             for name in ts.static_order():
                 if name not in graph:
                     continue
-                f.write("{} {} {}\n".format(name, types[name], ','.join(graph[name])))
+                f.write(f"{name} {types[name]} {','.join(graph[name])}\n")
 
     def package_id(self):
         self.info.requires["protobuf"].full_package_mode()
@@ -203,8 +187,8 @@ class GoogleAPIS(ConanFile):
     def package_info(self):
         with open(os.path.join(self.package_folder, "generated_targets.deps"), "r", encoding="utf-8") as f:
             for line in f.read().splitlines():
-                (name, type, deps) = line.rstrip('\n').split(' ')
+                (name, libtype, deps) = line.rstrip('\n').split(' ')
                 self.cpp_info.components[name].requires = deps.split(',')
-                if type == 'LIB':
+                if libtype == 'LIB':
                     self.cpp_info.components[name].libs = [name]
                 self.cpp_info.components[name].names["pkg_config"] = name

--- a/recipes/googleapis/all/conanfile.py
+++ b/recipes/googleapis/all/conanfile.py
@@ -186,7 +186,7 @@ class GoogleAPIS(ConanFile):
             graph[lib.cmake_target] = Node(mark=None, deps=lib.cmake_deps)
             types[lib.cmake_target] = 'LIB' if lib.srcs else 'INTERFACE'
 
-        def visit(name : str, dag : dict, L : list[str]):
+        def visit(name : str, dag : dict, L : list):
             # Ignore external deps that are not in the dag, such as protobuf::libprotobuf.
             # Also skip nodes that are already completely processed
             if name not in dag or dag[name].mark == 'DONE':

--- a/recipes/googleapis/all/patches/cci.20210730/001-fix-google-api-deps.patch
+++ b/recipes/googleapis/all/patches/cci.20210730/001-fix-google-api-deps.patch
@@ -1,0 +1,19 @@
+diff --git a/google/api/BUILD.bazel b/google/api/BUILD.bazel
+index ebb9ba8..7188218 100644
+--- a/google/api/BUILD.bazel
++++ b/google/api/BUILD.bazel
+@@ -509,12 +509,12 @@ cc_proto_library(
+ 
+ cc_proto_library(
+     name = "monitoring_cc_proto",
+-    deps = ["monitoring_proto"],
++    deps = [":monitoring_proto"],
+ )
+ 
+ cc_proto_library(
+     name = "quota_cc_proto",
+-    deps = ["quota_proto"],
++    deps = [":quota_proto"],
+ )
+ 
+ cc_proto_library(

--- a/recipes/googleapis/all/patches/cci.20211122/001-fix-google-api-deps.patch
+++ b/recipes/googleapis/all/patches/cci.20211122/001-fix-google-api-deps.patch
@@ -1,0 +1,19 @@
+diff --git a/google/api/BUILD.bazel b/google/api/BUILD.bazel
+index ebb9ba8..7188218 100644
+--- a/google/api/BUILD.bazel
++++ b/google/api/BUILD.bazel
+@@ -574,12 +574,12 @@ cc_proto_library(
+ 
+ cc_proto_library(
+     name = "monitoring_cc_proto",
+-    deps = ["monitoring_proto"],
++    deps = [":monitoring_proto"],
+ )
+ 
+ cc_proto_library(
+     name = "quota_cc_proto",
+-    deps = ["quota_proto"],
++    deps = [":quota_proto"],
+ )
+ 
+ cc_proto_library(

--- a/recipes/googleapis/all/patches/cci.20220531/001-fix-google-api-deps.patch
+++ b/recipes/googleapis/all/patches/cci.20220531/001-fix-google-api-deps.patch
@@ -1,0 +1,19 @@
+diff --git a/google/api/BUILD.bazel b/google/api/BUILD.bazel
+index ebb9ba8..7188218 100644
+--- a/google/api/BUILD.bazel
++++ b/google/api/BUILD.bazel
+@@ -578,12 +578,12 @@ cc_proto_library(
+ 
+ cc_proto_library(
+     name = "monitoring_cc_proto",
+-    deps = ["monitoring_proto"],
++    deps = [":monitoring_proto"],
+ )
+ 
+ cc_proto_library(
+     name = "quota_cc_proto",
+-    deps = ["quota_proto"],
++    deps = [":quota_proto"],
+ )
+ 
+ cc_proto_library(

--- a/recipes/googleapis/all/patches/cci.20220711/001-fix-google-api-deps.patch
+++ b/recipes/googleapis/all/patches/cci.20220711/001-fix-google-api-deps.patch
@@ -1,0 +1,19 @@
+diff --git a/google/api/BUILD.bazel b/google/api/BUILD.bazel
+index ebb9ba8..7188218 100644
+--- a/google/api/BUILD.bazel
++++ b/google/api/BUILD.bazel
+@@ -580,12 +580,12 @@ cc_proto_library(
+ 
+ cc_proto_library(
+     name = "monitoring_cc_proto",
+-    deps = ["monitoring_proto"],
++    deps = [":monitoring_proto"],
+ )
+ 
+ cc_proto_library(
+     name = "quota_cc_proto",
+-    deps = ["quota_proto"],
++    deps = [":quota_proto"],
+ )
+ 
+ cc_proto_library(

--- a/recipes/googleapis/all/test_package/conanfile.py
+++ b/recipes/googleapis/all/test_package/conanfile.py
@@ -7,7 +7,7 @@ from conan.tools.layout import cmake_layout
 
 class TestPackageConan(ConanFile):
     settings = "os", "arch", "compiler", "build_type"
-    generators = "CMakeDeps", "VirtualRunEnv"
+    generators = "cmake_find_package", "CMakeDeps", "VirtualRunEnv"
 
     def requirements(self):
         self.requires(self.tested_reference_str)

--- a/recipes/googleapis/all/test_package/test_package.cpp
+++ b/recipes/googleapis/all/test_package/test_package.cpp
@@ -1,14 +1,13 @@
 #include <iostream>
-#include "google/type/color.pb.h"
+#include <google/bigtable/v2/bigtable.pb.h>
 
 int main() {
     std::cout << "Conan - test package for googleapis\n";
 
-    google::type::Color c;
-    c.set_red(255);
-    c.set_blue(255);
+    google::bigtable::v2::CheckAndMutateRowRequest request;
+    request.set_table_name("projects/my-project/instances/my-instance/tables/my-table");
 
-    std::cout << c.DebugString();
+    std::cout << request.DebugString();
 
     return 0;
 }

